File: day_11.cpp
#include <iostream>
#include <vector>
#include <stdio.h>
#include <string>
#include <fstream>
#include <sstream>

uint64_t blink_stone(uint64_t stone, int blink_times, uint64_t stone_count)
{
    if (blink_times < 1)
        return stone_count;
    if (stone == 0)
    {
        stone_count = blink_stone(1, blink_times - 1, stone_count);
    }
    else
    {
        std::string split_stone = std::to_string(stone);
        if ((split_stone.length() % 2) == 0)
        {
            unsigned long stone1 = std::stol(split_stone.substr(0, split_stone.length() / 2));
            unsigned long stone2 = std::stol(split_stone.substr(split_stone.length() / 2));
            stone_count = blink_stone(stone1, blink_times - 1, stone_count) + blink_stone(stone2, blink_times - 1, stone_count);
        }
        else
        {
            stone_count = blink_stone(stone * 2024, blink_times - 1, stone_count);
        }
    }
    return stone_count;
}

void part(char part_no)
{
    std::cout << "Part " << part_no << "\n";
    std::string line, str_stone;
    std::ifstream file("day_11\\day_11_input.txt");
    uint64_t stones_count = 0;
    if (file.is_open()) 
    {

        std::getline(file, line);
        std::cout << line << "\n";
        std::vector<unsigned long> stones;
        std::stringstream str_stones(line);
        while (std::getline(str_stones, str_stone, ' '))
        {
            stones.push_back(std::stol(str_stone));
        }
        int blink_times = 25;
        if (part_no == '2')
        {
            blink_times = 75;
        }
        for (const uint64_t& stone : stones)
        {
            uint64_t count = blink_stone(stone, blink_times, 1);
            stones_count += count;
            std::cout << "Stone " << stone << " = " << count << std::endl;
        }
        std::cout << "Stones count = " << stones_count << "\n";
    }
    else
    {
        std::cout << "File not open!\n";
    }
    file.close();
}

int main()
{
    char part_no;
    std::cout << "Advent of Code - Day 11\n";
    std::cout << "Wich part do you want to test (1 or 2):";
    std::cin >> part_no;
    if (part_no == '1' || part_no == '2')
    {
        part(part_no);
    }
    else
    {
        std::cout << "Invalid part!\n";
    }
    std::cout << "Press ENTER to exit.";
    std::cin.clear();
    std::cin.ignore(std::numeric_limits <std::streamsize> ::max(), '\n');
    std::cin.get();
    //while (std::cin.get() != '\n');
    //std::cin >> part_no;
}

File: day_11.py
from functools import cache

def blink_stones(stones):
    new_stones = []
    for stone in stones:
        if stone == 0:
            new_stones.append(1)
        elif (len(str(stone)) % 2) == 0:
            split_stone = str(stone)
            split_stone = split_stone[0:len(split_stone)//2] + " " + split_stone[len(split_stone)//2:]
            split_stones = [int(s) for s in split_stone.split(" ")]
            new_stones.append(split_stones[0])
            new_stones.append(split_stones[1])
        else:
            new_stones.append(stone * 2024)
    return new_stones

@cache
def blink_stone(stone, blink_times, stone_count):
    if blink_times < 1:
        return stone_count
    if stone == 0:
        stone_count = blink_stone(1, blink_times - 1, stone_count)
    elif (len(str(stone)) % 2) == 0:
        split_stone = str(stone)
        split_stone = split_stone[0:len(split_stone)//2] + " " + split_stone[len(split_stone)//2:]
        split_stones = [int(s) for s in split_stone.split(" ")]
        stone_count = blink_stone(split_stones[0], blink_times - 1, stone_count) + blink_stone(split_stones[1], blink_times - 1, stone_count)
    else:
        stone_count = blink_stone(stone * 2024, blink_times - 1, stone_count)
    return stone_count

def part_one():
    stones_count = 0
    with open("./day_11/day_11_input.txt") as file:
        lines = [line.strip().split(" ") for line in file.readlines()]
        stones = [int(stone) for stone in lines[0]]
        print(stones)
        for stone in stones:
            stones_count += blink_stone(stone, 25, 1)
        #for blink in range(25):
        #    stones = blink_stones(stones)
            #print(stones)
        #stones_count = len(stones)
    print(f"Stone count after 25 blinks = {stones_count}")
    return stones_count

def part_two():
    stones_count = 0
    with open("./day_11/day_11_input.txt") as file:
        lines = [line.strip().split(" ") for line in file.readlines()]
        stones = [int(stone) for stone in lines[0]]
        print(stones)
        for stone in stones:
            count = blink_stone(stone, 75, 1)
            print(f"Blink stone {stone} array = {count}")
            stones_count += count
    print(f"Stone count after 75 blinks = {stones_count}")
    return stones_count

if __name__ == "__main__":
    print("Advent of Code - Day 11")
    part = input("Wich part do you want to test (1 or 2):")
    if part == "1":
        part_one()
    elif part == "2":
        part_two()
